RFC/src/RFC-0201_TariScript.md

```diff
@@ -0,0 +1,410 @@
+  \text{If the accounting is correct, all values will cancel} \\\\
+  \sum(k_i + \mathrm{H}(C_i \Vert s_i).G) - \sum(k_j + \mathrm{H}(C_j \Vert s_j).G)  \\\\
+  \text{The sum of all the blinding factors (times G) is the definition of the excess,}\\; x_s\cdot G = X_s \\\\
+  X_s + \sum(\mathrm{H}(C_i \Vert s_i).G) - \sum(\mathrm{H}(C_j \Vert s_j).G) \\\\
+  \text{Define}\\, \Lambda = \sum(\mathrm{H}(C_i \Vert s_i).G) - \sum(\mathrm{H}(C_j \Vert s_j).G) \\\\
+  X_s + \Lambda \\\\
+\end{array}  
+$$
+
+The kernels are unmodified, except that the excess will now include \\( \Lambda \\), representing the sum of all the
+commitments to the UTXO script hashes. This also means that the kernel signatures are calculated slightly differently:
+
+$$  
+\begin{array}
+  .s_i &= r_i + e.\bigl(k_i + \mathrm{H}(C_i \Vert s_i) \bigr) \\\\
+  s_i.G &= r_i.G + e.\bigl(k_i + \mathrm{H}(C_i \Vert s_i) \bigr).G \\\\
+  s_i.G &= R_i + e.\bigl(P_i + \mathrm{H}(C_i \Vert s_i)\bigr) \\\\
+\end{array}
+$$
+
+Summing the signatures, one can easily confirm that \\( X_s + \Lambda \\) signs the kernel correctly. The kernel offset
+is not included in this treatment, but it does not affect the result. One of the input commitments will be offset by a
+value selected by the sender and provided with the transaction data as usual. The signatures will still validate as
+usual, and the kernel offset will correct the overall excess balance.
+
+The same treatment extends to be block validation check. Note that since the individual kernel excesses can still be
+summed to obtain the overall block balance, the de-association of kernels and their outputs is maintained.
+
+### Checking the requirements
+
+Miners cannot modify the script hash, because it is committed to in the public excess value. Moreover, the \\( C_i \Vert
+s_i \\) pair is committed to, so miners can't, for example swap script hashes on commitments to keep the overall excess
+the same but still manipulate specific outputs.
+
+The UTXO creator(s) are also committing to the script hash by including it in the kernel signature.
+
+Thus all three requirements are satisfied and Tari Script, using this formulation should offer the same security
+guarantees that range proofs do.
+
+In particular, UTXOs can still be pruned because the \\( \Lambda \\) values change sign when used as inputs and will  
+cancel out in the overall balance in the same way that the pruned out excesses are.
+
+However, a problem arises now in that as it stands, the UTXOs _cannot_ be pruned because we would lose some data needed
+to verify the kernel signatures, i.e. \\( \mathrm{H}(C_i \Vert s_i) \\) and that data only exists in the UTXOs. However,
+we can salvage this situation fairly easily by noticing that we only need that _hash_ of the commitment and script hash.
+If we track an MMR of \\( C_i \Vert s_i \\), then those hashes are always available, even after the UTXOs themselves have
```

Comment:
Would it not be easier to commit to \\( C_i \Vert s_i \\) in the kernel it self than the header. 

if we add a mmr for \\( C_i \Vert s_i \\), data transfered by block would be:
+ 32 bytes for every utxo
+ 32 bytes for header. 

If we add it to the kernel it would be:
+ 32 bytes per tx. 
+ 32 bytes per utxo

So a block is a bit bigger. 

But storage is:
if we add a mmr for \\( C_i \Vert s_i \\), data stored is
 + 2^(k+1)-1  * 32 bytes where k = total number of utxos
If you just want to store the leaves its 
+ k* 32 bytres. 

If we add it to the kernel, storage is:
+ 32 bytes per kenel.

Reply:
I think we might get away to make this field optional as well. That way the extra data is only stored if its required

Reply:
Storage is definitely higher, but that's not great, but ok.

The problem with committing the hash in the kernel is that you permanently link kernels to outputs, breaking privacy which is not great at all. I saw the extra storage as the price we pay to maintain privacy.

I realised that fields can't really be optional, because then miners could go and remove everything, and nodes would allow it because missing data is valid.

Reply:
That can be locked up in the challenge removing the malleability.  But you're right, the privacy leak is big.